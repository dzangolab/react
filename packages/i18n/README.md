# Dzangolab React I18n

Adds internationalization to a react app.

## Requirements

* `@dzangolab/react-config: 0.1.0`

## Installation

```bash
pnpm install @dzangolab/react-config @dzangolab/react-i18n
```

## Usage

### Setup i18n

In `src/main.tsx`:

```ts
import i18n from "@dzangolab/react-i18n";
import i18n from "@dzangolab/react-i18n";
import React from "react";
import ReactDOM from "react-dom/client";

import App from "./App";
import config from "./config";


i18n(config.i18n);

...
```

### Resources

Create your resources under `src/locales/{locale}`.

For each locale, your resource should look like this:

```
{
  "locales": {
    "en": "English",
    "fr": "Français",
    "th": "ไทย"
  }
}
```

### Configuration

The package supports the same configuration options as the `react-i18next` package.

The configuration options must be defined under the `i18n` attribute of your app config (`AppConfig`).

```
# src/config.ts

const config: AppConfig = {
  ...
  i18n: {
    fallbackLng: "en",
    resources: {
      ...
    },
  },
  ...
};
```
