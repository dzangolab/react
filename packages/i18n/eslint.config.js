import path from "path";
import { fileURLToPath } from "url";

import { FlatCompat } from "@eslint/eslintrc";
import js from "@eslint/js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const compat = new FlatCompat({
  baseDirectory: __dirname,
  recommendedConfig: js.configs.recommended,
  allConfig: js.configs.all,
});

export default [
  {
    ignores: ["dist", "node_modules"],
  },
  ...compat.extends("@12deg/eslint-config/react"),
];
