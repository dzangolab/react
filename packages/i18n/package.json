{
  "name": "@dzangolab/react-i18n",
  "version": "0.8.5",
  "description": "React I18n Plugin",
  "exports": {
    ".": {
      "import": "./dist/DzangolabReactI18n.es.js",
      "require": "./dist/DzangolabReactI18n.umd.js"
    },
    "./dist/DzangolabReactI18n.css": "./dist/style.css"
  },
  "main": "dist//DzangolabReactI18n.umd.js",
  "module": "dist//DzangolabReactI18n.es.js",
  "types": "./dist/src/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "vite build && tsc --emitDeclarationOnly",
    "lint": "eslint . --ext .js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts",
    "lint:fix": "eslint . --ext .js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --fix",
    "sort-package": "npx sort-package-json",
    "test": "vitest --environment jsdom run --coverage",
    "test:component": "vitest --environment jsdom run component/",
    "test:snapshot": "vitest --environment jsdom run snapshot/",
    "test:unit": "vitest --environment jsdom run unit/",
    "test:watch": "vitest",
    "typecheck": "tsc --noEmit -p tsconfig.vitest.json --composite false"
  },
  "dependencies": {
    "@dzangolab/react-ui": "0.8.5",
    "i18next": "21.10.0",
    "i18next-browser-languagedetector": "6.1.8",
    "i18next-chained-backend": "^4.0.0",
    "i18next-resources-to-backend": "^1.0.0",
    "react-i18next": "11.18.6"
  },
  "devDependencies": {
    "@dzangolab/react-config": "0.8.5",
    "@testing-library/react": "14.0.0",
    "@types/i18n": "0.13.6",
    "@types/jsdom": "21.1.1",
    "@types/node": "18.11.18",
    "@types/react": "18.2.14",
    "@types/react-dom": "18.2.6",
    "@types/react-test-renderer": "18.0.0",
    "@vitejs/plugin-react": "3.1.0",
    "@vitest/coverage-c8": "0.30.1",
    "eslint": "8.42.0",
    "eslint-config-custom": "0.8.5",
    "eslint-plugin-react": "7.32.2",
    "prettier": "2.8.8",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-test-renderer": "18.2.0",
    "tsconfig": "0.8.5",
    "typescript": "5.1.3",
    "vite": "4.3.9",
    "vitest": "0.30.1"
  },
  "peerDependencies": {
    "@dzangolab/react-config": "0.8.5",
    "react": ">=18.2",
    "react-dom": ">=18.2"
  },
  "engines": {
    "node": ">=16",
    "pnpm": ">=8"
  }
}