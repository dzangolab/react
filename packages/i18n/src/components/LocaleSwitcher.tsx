import { DropdownMenuV2 as DropdownMenu } from "@dzangolab/react-ui";
import { MenuProps as PMenuProperties } from "primereact/menu";
import { useTranslation } from "react-i18next";

import "../css/locale-switcher.css";

export type LocalSwitcherProperties = PMenuProperties;

const LocaleSwitcher = ({ ...menuOptions }: LocalSwitcherProperties) => {
  const { i18n, t } = useTranslation("locales");

  const changeLocale = (newLocale: string) => {
    i18n.changeLanguage(newLocale);

    document.documentElement.lang = newLocale;
  };

  const locales =
    i18n.options.supportedLngs &&
    i18n.options.supportedLngs
      .filter((locale) => locale !== "cimode" && i18n.options.debug) // Filter out cimode from options. The mode shows key e.g. locale.english as value.
      .map((locale) => {
        return {
          command: () => changeLocale(locale),
          label: t(`locales.${locale}`),
        };
      });

  if (!locales || !(locales.length > 1)) {
    return null;
  }

  return (
    <span className="locale-switcher">
      <DropdownMenu
        menu={locales || []}
        menuLabel={t(`locales.${i18n.language}`)}
        {...menuOptions}
      />
    </span>
  );
};

export default LocaleSwitcher;
