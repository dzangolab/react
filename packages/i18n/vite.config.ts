import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

import react from "@vitejs/plugin-react";
import { defineConfig, loadEnv } from "vite";

import { dependencies, peerDependencies } from "./package.json";

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => {
  process.env = { ...process.env, ...loadEnv(mode, process.cwd()) };

  return {
    build: {
      lib: {
        entry: resolve(dirname(fileURLToPath(import.meta.url)), "src/index.ts"),
        fileName: (format) => `DzangolabReactI18n.${format}.js`,
        name: "@dzangolab/react-i18n",
      },
      rollupOptions: {
        external: [
          ...Object.keys(peerDependencies),
          ...Object.keys(dependencies),
        ],
        output: {
          exports: "named",
          globals: {
            "@dzangolab/react-config": "DzangolabReactConfig",
            i18next: "I18next",
            "i18next-browser-languagedetector":
              "I18nextBrowserLanguagedetector",
            "i18next-chained-backend": "I18nextChainedBackend",
            "i18next-resources-to-backend": "I18nextResourcesToBackend",
            react: "React",
            "react-dom": "ReactDom",
            "react-i18next": "ReactI18next",
            "react-outside-click-handler": "ReactOutsideClickHandler",
          },
        },
      },
      target: "esnext",
      sourcemap: true,
    },
    plugins: [react()],
    resolve: {
      alias: {
        "@": fileURLToPath(new URL("./src", import.meta.url)),
      },
    },
    server: {
      port: Number(process.env.VITE_APP_PORT) || 8889,
    },
    optimizeDeps: {
      include: ["react/jsx-runtime"],
    },
  };
});
