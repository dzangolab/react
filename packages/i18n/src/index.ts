import { use } from "i18next";
import LanguageDetector from "i18next-browser-languagedetector";
import ChainedBackend from "i18next-chained-backend";
import resourcesToBackend from "i18next-resources-to-backend";
import { initReactI18next, useTranslation, Trans } from "react-i18next";

import LocaleSwitcher from "./components/LocaleSwitcher";

/* eslint-disable-next-line @typescript-eslint/no-unused-vars */
import type { AppConfig } from "@prefabs.tech/react-config";
import type { InitOptions } from "i18next";

const plugin = (options: InitOptions) => {
  use(LanguageDetector)
    .use(initReactI18next)
    .use(ChainedBackend)
    .init({
      ...options,
      interpolation: {
        escapeValue: false,
      },
      backend: {
        backends: [
          resourcesToBackend((language, namespace, callback) => {
            import(`./locales/${language}/${namespace}.json`)
              .then((resources) => {
                callback(null, resources);
              })
              .catch((error) => {
                callback(error, null);
              });
          }),
        ],
      },
    });
};

declare module "@prefabs.tech/react-config" {
  export interface AppConfig {
    i18n: InitOptions;
  }
}

export default plugin;

export { LocaleSwitcher, useTranslation, Trans };
