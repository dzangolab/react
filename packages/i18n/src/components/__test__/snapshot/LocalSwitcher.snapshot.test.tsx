import { create } from "react-test-renderer";
import { expect, test } from "vitest";

import i18n from "../../../index";
import LocaleSwitcher from "../../LocaleSwitcher";

import type {
  ReactTestRenderer,
  ReactTestRendererJSON,
} from "react-test-renderer";

function toJson(component: ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as ReactTestRendererJSON;
}

test("Component matches snapshot", () => {
  const i18nConfig = {
    appendNamespaceToCIMode: true,
    debug: true,
    defaultNS: "app",
    fallbackLng: "en",
    lng: "en",
    resources: {
      en: {},
      fr: {},
    },
    supportedLngs: ["en", "fr"],
    react: {
      useSuspense: false,
    },
    locales: "en:English,fr:Fran√ßais",
  };

  i18n(i18nConfig);
  const component = create(<LocaleSwitcher />);
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
