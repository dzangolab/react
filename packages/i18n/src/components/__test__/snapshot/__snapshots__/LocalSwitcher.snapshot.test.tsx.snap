// Vitest Snapshot v1

exports[`Component matches snapshot 1`] = `
<div>
  <nav
    className="locale-switcher "
  >
    <div
      className="locale truncated"
      onClick={[Function]}
    >
      locales.en
      <span
        className="truncated "
      >
        â–¾
      </span>
    </div>
  </nav>
</div>
`;
